src/Databases/DatabaseFilesystem.cpp

```diff
@@ -0,0 +1,189 @@
+#include <Databases/DatabaseFilesystem.h>
+
+#include <IO/Operators.h>
+#include <IO/WriteBufferFromString.h>
+#include <Interpreters/Context.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ParserCreateQuery.h>
+#include <Parsers/parseQuery.h>
+#include <Storages/IStorage.h>
+#include <TableFunctions/TableFunctionFactory.h>
+
+#include <filesystem>
+
+namespace fs = std::filesystem;
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int UNKNOWN_TABLE;
+    extern const int DATABASE_ACCESS_DENIED;
+    extern const int BAD_ARGUMENTS;
+}
+
+DatabaseFilesystem::DatabaseFilesystem(const String & name_, const String & path_, ContextPtr context_)
+    : IDatabase(name_), WithContext(context_->getGlobalContext()), path(path_), log(&Poco::Logger::get("DatabaseFileSystem(" + name_ + ")"))
+{
+    fs::path user_files_path;
+    if (context_->getApplicationType() != Context::ApplicationType::LOCAL)
+        user_files_path = fs::canonical(fs::path(getContext()->getUserFilesPath()));
+
+    if (fs::path(path).is_relative())
+        path = user_files_path / path;
+
+    path = fs::absolute(path).lexically_normal().string();
```

Comment:
hm, should we also add a check that path actually exists?

Reply:
It seems to me that this is redundant. There is a check for existence in the `tryGetTable` and inside `TableFunctionFile`. In addition, you can use relative paths that will make the original path correct. For example, `path = ..../user_files/unknown_dir`, and you cat read table `../known_dir/table.csv`

Reply:
fixed, non-existent paths are useless, I interpret them as BAD_ARGUMENTS